File: src/reader/ReaderAdapter.java
package reader;

import readers.RFIDReader;
import readers.ThingMagicM7EReader;
import readers.GenericLLRPReader;
import java.util.Timer;
import java.util.TimerTask;

/**
 * Adaptador principal que delega a implementaciones específicas de lectores
 * Arquitectura modular: el agente LLRP no sabe qué lector físico está usando
 */
public class ReaderAdapter {

    private RFIDReader realReader;
    private Timer inventoryTimer;
    private boolean inventoryRunning = false;

    public ReaderAdapter() {
        // Selección dinámica de reader:
        // 1) Si se define la propiedad del sistema `llrp.reader.type`, se usa esa.
        // 2) Si no, por defecto se usa el modo mock (permite ejecutar el agente sin hardware).
        String requested = System.getProperty("llrp.reader.type");
        if (requested == null || requested.isEmpty()) {
            requested = System.getenv("LLRP_READER_TYPE");
        }

        if (requested == null || requested.isEmpty()) {
            // Por defecto intentamos usar un reader LLRP genérico (para lectores que implementan LLRP)
            requested = "llrp";
        }

        setReaderType(requested);
    }

    /**
     * Método para cambiar dinámicamente el tipo de reader
     */
    public void setReaderType(String readerType) {
        // Desconectar reader actual
        if (realReader != null) {
            try {
                realReader.disconnect();
            } catch (Exception ignored) {
            }
        }

        // Crear nuevo reader según tipo
        RFIDReader candidate;
        switch (readerType.toLowerCase()) {
            case "thingmagic":
            case "m7e":
                candidate = new ThingMagicM7EReader();
                break;
            case "llrp":
                candidate = new GenericLLRPReader();
                break;
            case "mock":
            default:
                candidate = new MockRFIDReader();
                break;
        }

        // Intentar conectar; si falla, regresar a mock para que el agente siga funcionando
        try {
            candidate.connect();
            realReader = candidate;
        } catch (Exception e) {
            System.err.println("Could not connect using reader type '" + readerType + "': " + e.getMessage());
            System.err.println("Falling back to mock reader (no hardware connection)");
            realReader = new MockRFIDReader();
            try {
                realReader.connect();
            } catch (Exception ignored) {
                // no-op
            }
        }
    }

    public void startInventory() {
        try {
            if (realReader != null) {
                realReader.startInventory();
                inventoryRunning = true;

                // Iniciar timer para mostrar tags periódicamente
                inventoryTimer = new Timer();
                inventoryTimer.scheduleAtFixedRate(new TimerTask() {
                    @Override
                    public void run() {
                        String[] tags = readTags();
                        if (tags.length > 0) {
                            System.out.println("Tags detected: " + java.util.Arrays.toString(tags));
                        }
                    }
                }, 0, 2000); // cada 2 segundos
            }
        } catch (Exception e) {
            System.err.println("Failed to start inventory: " + e.getMessage());
        }
    }

    public void stopInventory() {
        try {
            if (realReader != null) {
                realReader.stopInventory();
                inventoryRunning = false;
            }

            if (inventoryTimer != null) {
                inventoryTimer.cancel();
                inventoryTimer = null;
            }
        } catch (Exception e) {
            System.err.println("Failed to stop inventory: " + e.getMessage());
        }
    }

    public String[] readTags() {
        if (realReader != null) {
            try {
                return realReader.readTags();
            } catch (Exception e) {
                System.err.println("Failed to read tags: " + e.getMessage());
                return new String[0];
            }
        }
        return new String[0];
    }

    public boolean isInventoryRunning() {
        return inventoryRunning && (realReader != null ? realReader.isInventoryRunning() : false);
    }

    /**
     * Implementación mock para cuando no hay lector real disponible
     */
    private class MockRFIDReader implements RFIDReader {
        @Override
        public void connect() throws Exception {
            System.out.println("Mock reader connected (no real hardware)");
        }

        @Override
        public void disconnect() {
            System.out.println("Mock reader disconnected");
        }

        @Override
        public void startInventory() throws Exception {
            System.out.println("Mock inventory started");
        }

        @Override
        public void stopInventory() throws Exception {
            System.out.println("Mock inventory stopped");
        }

        @Override
        public String[] readTags() {
            return new String[]{
                "E2000017221101441890B3AA",
                "E2000017221101441890B3AB"
            };
        }

        @Override
        public boolean isInventoryRunning() {
            return true;
        }

        @Override
        public void configureReader() throws Exception {
            System.out.println("Mock reader configured");
        }
    }
}

File: cpp/src/reader/ReaderAdapter.cpp
#include "ReaderAdapter.h"
#include "../readers/GenericLLRPReader.h"
#include "../readers/MockLLRPReader.h"

#ifdef LLRP_AGENT_ENABLE_M7E
#include "../readers/ThingMagicM7EReader.h"
#endif

#include <algorithm>
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <stdexcept>

namespace reader {

// ────────────────────────────────────────────────────────────────────────────
// Constructor / Destructor
// ────────────────────────────────────────────────────────────────────────────

ReaderAdapter::ReaderAdapter() {
    // 1) entorno, 2) defecto "llrp"
    std::string type;
    if (const char *e = std::getenv("LLRP_READER_TYPE")) type = e;
    if (type.empty()) type = "llrp";

    // Convertir a minúsculas
    std::transform(type.begin(), type.end(), type.begin(), ::tolower);

    setReaderType(type);
}

ReaderAdapter::~ReaderAdapter() {
    stopPeriodicTagPrint();
    if (realReader_) {
        try { realReader_->disconnect(); } catch (...) {}
    }
}

// ────────────────────────────────────────────────────────────────────────────
// Selección dinámica de reader
// ────────────────────────────────────────────────────────────────────────────

void ReaderAdapter::setReaderType(const std::string &type) {
    // Desconectar el actual
    if (realReader_) {
        try { realReader_->disconnect(); } catch (...) {}
        realReader_.reset();
    }

    std::unique_ptr<readers::RFIDReader> candidate;
    if (type == "llrp")
        candidate = std::make_unique<readers::GenericLLRPReader>();
#ifdef LLRP_AGENT_ENABLE_M7E
    else if (type == "thingmagic" || type == "m7e")
        candidate = std::make_unique<readers::ThingMagicM7EReader>();
#endif
    else
        throw std::invalid_argument("Unknown reader type: " + type);

    try {
        candidate->connect();
        realReader_ = std::move(candidate);
        std::cout << "[ReaderAdapter] Successfully connected using reader type '" << type << "'\n";
    } catch (const std::exception &e) {
        std::cerr << "Could not connect using reader type '" << type << "': " << e.what() << "\n";
        
        // Fallback: usar MockLLRPReader solo si se solicita explícitamente
        const char *useMockEnv = std::getenv("LLRP_USE_MOCK_IF_UNAVAILABLE");
        bool useMockFallback = useMockEnv && (std::string(useMockEnv) == "1" || std::string(useMockEnv) == "true");
        
        if (useMockFallback) {
            std::cout << "[ReaderAdapter] LLRP_USE_MOCK_IF_UNAVAILABLE=true, falling back to MockLLRPReader\n";
            try {
                auto mockCandidate = std::make_unique<readers::MockLLRPReader>();
                mockCandidate->connect();
                realReader_ = std::move(mockCandidate);
                std::cout << "[ReaderAdapter] Mock reader initialized successfully\n";
            } catch (const std::exception &mockE) {
                std::cerr << "Failed to initialize MockLLRPReader: " << mockE.what() << "\n";
                realReader_.reset();
            }
        } else {
            std::cout << "[ReaderAdapter] No reader available (set LLRP_USE_MOCK_IF_UNAVAILABLE=1 to use mock)\n";
            realReader_.reset();
        }
    }
}

// ────────────────────────────────────────────────────────────────────────────
// Control de inventario
// ────────────────────────────────────────────────────────────────────────────

void ReaderAdapter::startInventory() {
    if (inventoryRunning_) {
        return;
    }
    if (!realReader_) return;
    try {
        realReader_->startInventory();
        inventoryRunning_ = true;
        startPeriodicTagPrint();
    } catch (const std::exception &e) {
        std::cerr << "Failed to start inventory: " << e.what() << "\n";
    }
}

void ReaderAdapter::stopInventory() {
    stopPeriodicTagPrint();
    if (!realReader_) return;
    try {
        realReader_->stopInventory();
        inventoryRunning_ = false;
    } catch (const std::exception &e) {
        std::cerr << "Failed to stop inventory: " << e.what() << "\n";
    }
}

std::vector<std::string> ReaderAdapter::readTags() {
    if (!realReader_) return {};
    try {
        return realReader_->readTags();
    } catch (const std::exception &e) {
        std::cerr << "Failed to read tags: " << e.what() << "\n";
        return {};
    }
}

bool ReaderAdapter::isInventoryRunning() const {
    return inventoryRunning_;
}

// ────────────────────────────────────────────────────────────────────────────
// Timer periódico (equivalente a java.util.Timer)
// ────────────────────────────────────────────────────────────────────────────

void ReaderAdapter::startPeriodicTagPrint() {
    if (timerThread_.joinable()) {
        return;
    }
    timerRunning_ = true;
    timerThread_ = std::thread([this]() {
        while (timerRunning_) {
            std::this_thread::sleep_for(std::chrono::seconds(2));
            if (!timerRunning_) break;
            auto tags = readTags();
            if (!tags.empty()) {
                std::cout << "Tags detected: [";
                for (size_t i = 0; i < tags.size(); ++i) {
                    if (i) std::cout << ", ";
                    std::cout << tags[i];
                }
                std::cout << "]\n";
            }
        }
    });
}

void ReaderAdapter::stopPeriodicTagPrint() {
    timerRunning_ = false;
    if (timerThread_.joinable()) timerThread_.join();
}

} // namespace reader
